src/pstl_offload.cpp

```diff
@@ -96,59 +97,311 @@ __original_msize(void* __user_ptr)
     return __orig_msize(__user_ptr);
 }
 
+
+static auto
+__get_original_realloc()
+{
+    using __realloc_func_type = void* (*)(void*, std::size_t);
+
+    static __realloc_func_type __orig_realloc = __realloc_func_type(dlsym(RTLD_NEXT, "realloc"));
+    return __orig_realloc;
+}
+
+inline bool
+__is_ptr_page_aligned(void* __p)
+{
+    return (uintptr_t)__p % __get_memory_page_size() == 0;
+}
+
+struct __hash_aligned_ptr
+{
+    uintptr_t operator()(void* __p) const
+    {
+        // We know that addresses are at least page-aligned, so, expecting page at least
+        // 4K-aligned, drop 11 right bits that are zeros, and treat rest as a pointer,
+        // hoping that an underlying Standard Library support this well.
+        static constexpr unsigned __ptr_shift = 11;
+        // current page size is at least 4K
+        assert(__get_memory_page_size() >= 1 << __ptr_shift);
```

Comment:
```suggestion
        assert(__get_memory_page_size() >= (1 << __ptr_shift));
```

I understand that this is equivalent to what is in the PR but I would add parentheses for the sake of readability (without remembering operators priority)

Reply:
Done.